src/middleware/compression.rs

```diff
@@ -86,44 +86,48 @@ impl Compression {
 
     /// Consumes the response and returns an encoded version of it.
     fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
-        match encoding {
-            Encoding::Gzip => {
-                let mut body = Body::empty();
-                std::mem::swap(&mut body, res.body_mut());
-                let stream = stream::GzipEncoder::new(body, self.settings.gzip);
-                *res.body_mut() = Body::from_stream(stream);
-                res.headers_mut()
-                    .append(CONTENT_ENCODING, encoding.to_header_value());
-                res
-            }
-            Encoding::Deflate => {
-                let mut body = Body::empty();
-                std::mem::swap(&mut body, res.body_mut());
-                let stream = stream::ZlibEncoder::new(body, self.settings.deflate);
-                *res.body_mut() = Body::from_stream(stream);
-                res.headers_mut()
-                    .append(CONTENT_ENCODING, encoding.to_header_value());
-                res
-            }
-            Encoding::Brotli => {
-                let mut body = Body::empty();
-                std::mem::swap(&mut body, res.body_mut());
-                let stream = stream::BrotliEncoder::new(body, self.settings.brotli);
-                *res.body_mut() = Body::from_stream(stream);
-                res.headers_mut()
-                    .append(CONTENT_ENCODING, encoding.to_header_value());
-                res
-            }
-            Encoding::Zstd => {
-                let mut body = Body::empty();
-                std::mem::swap(&mut body, res.body_mut());
-                let stream = stream::ZstdEncoder::new(body, self.settings.zstd);
-                *res.body_mut() = Body::from_stream(stream);
-                res.headers_mut()
-                    .append(CONTENT_ENCODING, encoding.to_header_value());
-                res
+        if res.headers().get(CONTENT_ENCODING).is_some() {
+            res // avoid double-encoding a given response
```

Comment:
Should we do a quick return here so that the else branch below could have one indent lesser?

Reply:
We could. I'll try it in a sec, it might look better.

Although I'm unsure if the borrow checker will complain without the else block anyway... Will try in a moment